common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+        : SocketAbstractImpl(socket(AF_PACKET, SOCK_RAW, htons(protocol)))
+    {
+    }
+};
+
+class UdpSocket : public SocketAbstractImpl {
+public:
+    UdpSocket() : SocketAbstractImpl(socket(AF_INET, SOCK_DGRAM, IPPROTO_IP)) {}
+};
+
+class TcpSocket : public SocketAbstractImpl {
+public:
+    TcpSocket() : SocketAbstractImpl(socket(AF_INET, SOCK_STREAM, IPPROTO_IP)) {}
+};
+
+class UdsSocket : public SocketAbstractImpl {
+public:
+    explicit UdsSocket() : SocketAbstractImpl(socket(AF_UNIX, SOCK_STREAM, 0)) {}
+};
+
+class AbstractNetlinkSocket : public SocketAbstractImpl {
+protected:
+    explicit AbstractNetlinkSocket(uint16_t protocol)
+        : SocketAbstractImpl(socket(AF_NETLINK, SOCK_RAW, protocol))
+    {
+    }
+};
+
+class NetlinkRouteSocket : public AbstractNetlinkSocket {
+public:
+    NetlinkRouteSocket() : AbstractNetlinkSocket(NETLINK_ROUTE) {}
+};
+
+/**
+ * This class implements the Connection interface with methods that wrap the recv(), recvfrom and
+ * recvmsg() system calls.
+ */
+class SocketConnectionImpl : public Connection {
+public:
+    explicit SocketConnectionImpl(const std::shared_ptr<Socket> &socket) : m_socket(socket) {}
+
+    std::shared_ptr<Socket> socket() override { return m_socket; }
+    int receive(Buffer &buffer) override
+    {
+        return ::recv(m_socket->fd(), buffer.data(), buffer.size(), MSG_DONTWAIT);
+    }
+    int receive_from(Buffer &buffer, SockAddr &address) override
+    {
+        return ::recvfrom(m_socket->fd(), buffer.data(), buffer.size(), MSG_DONTWAIT,
+                          address.sockaddr(), &address.length());
+    }
+    int receive_message(MsgHdr &message) override
+    {
+        // TODO
+        return -1;
+    }
+    int send(const Buffer &buffer, size_t length) override
+    {
+        return ::send(m_socket->fd(), buffer.data(), length, MSG_NOSIGNAL);
+    }
+    int send_to(const Buffer &buffer, size_t length, const SockAddr &address) override
+    {
+        return ::sendto(m_socket->fd(), buffer.data(), buffer.size(), 0, address.sockaddr(),
+                        address.length());
+    }
+    int send_message(const MsgHdr &message) override
+    {
+        // TODO
+        return -1;
+    }
+
+private:
+    /**
+     * Connected socket used by this connection object.
+     */
+    std::shared_ptr<Socket> m_socket;
+};
+
+class ServerSocketAbstractImpl : public ServerSocket {
+public:
+    int bind(const SockAddr &address)
+    {
+        return ::bind(m_socket->fd(), address.sockaddr(), address.length());
+    }
+
+    int listen(int backlog) { return ::listen(m_socket->fd(), backlog); }
+
+    std::unique_ptr<Connection> accept(SockAddr &address) override
+    {
+        int s = ::accept(m_socket->fd(), address.sockaddr(), &address.length());
+        if (FileDescriptor::invalid_descriptor == s) {
+            return nullptr;
+        }
+
+        return std::make_unique<SocketConnectionImpl>(std::make_shared<ConnectedSocket>(s));
+    }

```

Comment:
I agree. Suggested design fits perfectly in the scenario you've just described

Reply:
I forgot to write the point I wanted to make :-)

... Therefore, I don't think it's useful to have separate socket()/bind()/listen()/accept() wrappers. Instead, I think it is a lot more useful to have a class that implements the scenario that we need, rather than simply rewrite the C functions.

Reply:
Interface `ServerSocket` and these functions are required if we do want to test the classes that use sockets. _This is the way_